### DISCUSSION TITLE
Proposed extensions to the schema specification

### DISCUSSION THREAD

**Discussion by larsgw** [kdl-org/kdl#107] [Ideas]
1. A top-level `metadata` (or `info`) node. Authors, title/description, URLs (documentation, persistent self-URL for linked data), publication dates and version information.
  2. A top-level `definitions` node. In the example schema, the props `description`, `ref` and `id` are shared among various nodes throughout the documentation. To remove the duplication one could give the first definition an `id` and `ref` it in other places but the first definition is not necessarily the most natural place to put it. A `definitions` node could house `node`, `prop`, `value` and `children` nodes.

(Edit: I see there are `description` props already on `document`, but I think more metadata is still helpful. The same goes for `schema-url` if it's a reference to the current schema; if it's a reference to the schema schema, like an XML namespace, then a top-level `schema` node to contain `metadata`, `definitions` and `document` might be good too.)

**Comment by zkat** [comment]
I like both of these ideas and will gladly take PRs for them!!

**Reply by larsgw** [reply]
Great! What do you want to do with `document`s `description` and `schema-url`? Just move them there?

**Reply by zkat** [reply]
yeah. Can we also put the `definitions` and `metadata` nodes _inside_ `document`?

Thinking about it... I would like to keep `description` at the `document` level because it's handy to have a human description up there? What do you think? Schema-url can move, tho.

**Reply by larsgw** [reply]
> yeah. Can we also put the `definitions` and `metadata` nodes _inside_ `document`?

Of course, I thought `document` still only had the root `node`s but I see there is other configuration there too now.

> Thinking about it... I would like to keep `description` at the `document` level because it's handy to have a human description up there? What do you think? Schema-url can move, tho.

What about something like this:

```kdl
schema {
    metadata {
        description "KDL Schema KDL schema in KDL" lang="en"
        description "KDL Schema KDL schema in KDL" lang="nl"

        author "Kat Marchán" {
            link "https://github.com/zkat" rel="self" type="github"
        }
        author "Lars Willighagen" // or `contributor` in this case if that should be distinguished

        link "https://github.com/zkat/kdl" rel="self"
    }
    options {
        other-nodes-allowed true
    }
    definitions {
        children id="#validations" {
            node "format" // ...
            // etc.
        }
    }
    document {
        node "document" {
            min 1
            max 1
            prop "schema-url" description="URL where you can find this schema. Informational only." {
                type "url"
            }
            prop "description" description="General purpose and description for this document schema." {
                type "string"
            }
            children id="node-children" {
                node "node-names" description="Validations to apply specifically to arbitrary node names" {
                    children ref="#validations"
                }
                node "other-nodes-allowed" description="Whether to allow child nodes other than the ones explicitly listed. Defaults to 'false'." {
                    max 1
                    value {
                        min 1
                        max 1
                        type "boolean"
                    }
                }
                // etc.
            }
        }
    }
}
```

**Reply by larsgw** [reply]
It might depend on experience but I think I would personally be more drawn to the metadata section to see the description than the first line (especially coming from XML and HTML).

**Reply by zkat** [reply]
I'm more inclined to put everything _inside_ `document`:

```kdl
document {
    metadata {
        ...
    }
    definitions {
        ...
    }
    other-nodes-allowed true
    node "document" { ... }
}
```

**Reply by larsgw** [reply]
I like having a node for the KDL document root, especially since it can contain multiple `node` nodes (and those shouldn't have a `other-nodes-allowed` or `metadata` in between). The other things also all say something about the schema: what it is (`metadata`), how it should be applied (`other-nodes-allowed` etc.). In my opinion that is distinct from `document`, which directly says something about the files that the schema should be applied to (that is how I interpreted it at least). I am not entirely sure where `definitions` fits in this though.

**Reply by zkat** [reply]
what you're using as `schema` is literally what `document` is meant to be, though. The `document` node can be renamed to `schema` and we'd basically have the same behavior.

**Reply by larsgw** [reply]
Ah, okay. My bad.

**Comment by zkat** [comment]
I'm converting this to an issue because I want to make sure it's _resolved_ for #112
